### DISCUSSION TITLE
Level Zero - Best Practices and Guidelines

### DISCUSSION THREAD

**Discussion by jjfumero** [intel/compute-runtime#520] [Ideas]
Hi all, 
   I am not sure if this exists already but is there any guideline about best practices with Level Zero?

I think it will be good to have Level Zero guidelines for:
a) Programming for high-performance kernels 
b) Programming for low energy consumption
c) Programming for good memory management (not necessarily for high-performance kernels)

For example, perhaps using low-latency command lists can increase overall performance. Also, which flags are the most critical to achieving performance, etc. IMO the Level Zero community will appreciate this, and it will help to get wider adoption of this technology.

Issue linked #515

**Comment by jandres742** [comment]
Hi @jjfumero 
Please see here for the effort we have on our repo https://github.com/intel/compute-runtime/tree/master/programmers-guide. We are aiming at having there any information that may be useful for users to achieve high and well-behaved programs with L0 and OCL.

Now, generally speaking, our design aims at having everything working out of the box. The debug keys we have are usually for debug, as well as for experimenting with a feature while we try it before making it default.

In your question for instance about which flags allow low-latency command lists to achieve high performance, our objective is that they work with the default build, w/o user needing to use any debug key.  For low-latency lists, we had `EnableFlushTaskSubmission` (https://github.com/intel/compute-runtime/blob/3490b489ad206eb647f12870da39bfecf30914f3/shared/source/debug_settings/debug_variables_base.inl#L372) to achieve high performance, but we have already set that by default since 22bbce42dd25aa5a4a5cde970f975940cfd8db23. 

Additional resources in the oneAPI umbrella include:

https://www.intel.com/content/dam/develop/external/us/en/documents/oneapi-gpu-optimization-guide.pdf

https://www.intel.com/content/www/us/en/develop/documentation/oneapi-programming-guide/top/introduction-to-oneapi-programming.html
